### ISSUE TITLE
Thanos returns null result when using the unless operator and a non-empty RHS vector

### ISSUE THREAD

**Issue by roganartu** [improbable-eng/thanos#576]
**Thanos, Prometheus and Golang version used**

```
thanos, version 0.1.0-rc.2 (branch: HEAD, revision: 53e4d69ebba12baef46543468992619fcfd524ef)
  build user:       root@c7199d758b5e
  build date:       20180705-12:54:50
  go version:       go1.10.3
```

```
prometheus, version 2.3.1 (branch: stable, revision: aa4d706697f918d18815214db2823649ffc95626)
  build user:       atl@shbam1
  build date:       20180711-18:12:32
  go version:       go1.10.1
```
(this is a fork of prom, but afaict it doesn't impact the bug)

**What happened**

When issuing a query that uses the `unless` binary operator and the RHS of the unless is non-empty, Thanos returns `null` as the result instead of `[]`, which causes the UI to continue displaying a loading icon even though the query already returned.

**What you expected to happen**

Thanos to return `[]` as the result and the result window to display `no data`.

**How to reproduce it (as minimally and precisely as possible)**:

Issue the following query via Thanos:

```
vector(1) > 0 unless vector(1)
```

The response will be:

```
{
   "status": "success",
   "data": {
      "resultType": "vector",
      "result": null
   }
}
```
If you issued the query via Thanos UI you will also notice the loading gif continues to spin indefinitely after the query request has returned. There will be a javascript error in the console:

```
TypeError: data.result is null graph.js:772:1
```

If you query Prometheus directly it will be:

```
{
   "status": "success",
   "data": {
      "resultType": "vector",
      "result": []
   }
}
```
It also returns an empty list if you query Thanos with a modified single part of the query that would be expected to return no data, eg: `vector(0) > 0`.

**Anything else we need to know**

I am running Thanos as a query proxy in front of multiple Prometheus nodes and using the dedup feature. I tried deselecting "deduplication" in the Thanos UI and am still able to reproduce however. I wouldn't really expect the given query to hit more than one Prometheus node anyway, since it doesn't touch any metrics.

**Comment by bwplotka** [comment]
Thanks for this input and all details. Sounds like a valid bug, but it is worth to say that we use exactly same code for PromQL evaluation as Prometheus does. 

Worth to dive deeper are try to track this down with unit tests on thanos querier. Maybe something that was recently fixed on Prometheus web API side? To me it sounds like fix on response json serialization  than promQL (:

**Comment by roganartu** [comment]
I did some more digging since you said Thanos uses the same query endpoint logic as Prometheus. Turns out this is fixed in v0.1.0, specifically by 91b8391 which changed the prom version in Gopkg.lock: https://github.com/improbable-eng/thanos/commit/91b8391#diff-bd247e83efc3c45ae9e8c47233249f18R348

There's only a handful of prom changes in between those hashes, but the return result changes from `null` to `[]` in: https://github.com/prometheus/prometheus/commit/dd6781add2390eb2f2251e108f2ef835eba44b05 which is a huge change, so it seems this was an old prom behaviour that was fixed either intentionally or by accident.

I'm going to close this, thanks for nudging me in the right direction.
